File: scripts/generate_vendor_sbom.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const manifest = JSON.parse(
  fs.readFileSync(path.join(ROOT, "views", "vendor", "vendor-manifest.json"), "utf8"),
);
const components = manifest.assets.flatMap((asset) => Object.entries(asset.files).map(
  ([file, hash]) => ({
    type: "file",
    name: file,
    group: asset.name,
    version: asset.version,
    licenses: [{ license: { id: asset.license } }],
    hashes: [{ alg: "SHA-256", content: hash }],
    externalReferences: [
      { type: "distribution", url: asset.source },
      ...(asset.updateSource ? [{ type: "documentation", url: asset.updateSource }] : []),
    ],
  }),
));
const bom = {
  bomFormat: "CycloneDX",
  specVersion: "1.6",
  version: 1,
  metadata: {
    component: { type: "application", name: "BiddingFlow vendored browser assets", version: "1.0.0" },
  },
  components,
};
const outputDirectory = path.join(ROOT, "sbom");
fs.mkdirSync(outputDirectory, { recursive: true });
fs.writeFileSync(
  path.join(outputDirectory, "vendor.cdx.json"),
  `${JSON.stringify(bom, null, 2)}\n`,
  "utf8",
);
console.log(`Vendor SBOM generated (${components.length} components).`);

File: scripts/generate_vendor_sbom.py
"""Generate a CycloneDX SBOM for pinned vendored browser assets."""

from __future__ import annotations

import argparse
import json
from datetime import datetime, timezone
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
MANIFEST = ROOT / "views" / "vendor" / "vendor-manifest.json"


def build_sbom() -> dict:
    source = json.loads(MANIFEST.read_text(encoding="utf-8"))
    components = []
    for asset in source["assets"]:
        hashes = [
            {"alg": "SHA-256", "content": digest}
            for digest in sorted(asset["files"].values())
        ]
        component = {
            "type": "library",
            "name": asset["name"],
            "version": asset["version"],
            "hashes": hashes,
            "externalReferences": [
                {"type": "distribution", "url": asset["source"]}
            ],
        }
        if asset.get("license"):
            component["licenses"] = [
                {"license": {"id": asset["license"]}}
            ]
        components.append(component)
    return {
        "bomFormat": "CycloneDX",
        "specVersion": "1.6",
        "serialNumber": "urn:uuid:00000000-0000-4000-8000-000000000001",
        "version": 1,
        "metadata": {
            "timestamp": datetime.now(timezone.utc)
            .isoformat()
            .replace("+00:00", "Z"),
            "component": {
                "type": "application",
                "name": "biddingflow-vendored-browser-assets",
                "version": "1.0.0",
            },
        },
        "components": components,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--output",
        default="release/evidence/vendor-sbom.cdx.json",
    )
    args = parser.parse_args()
    output = (ROOT / args.output).resolve()
    output.relative_to(ROOT)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(
        json.dumps(build_sbom(), ensure_ascii=False, indent=2, sort_keys=True)
        + "\n",
        encoding="utf-8",
    )
    print(f"Vendor SBOM written to {output.relative_to(ROOT).as_posix()}.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
